### PR TITLE
refactor(v2): convert `@docusaurus/plugin-ideal-image` to TypeScript

### PR DESCRIPTION
**Pull request by gimdongwoo**
refactor @docusaurus/plugin-ideal-image

## Motivation

I want to refactor every plugin to typescript. It helps consistency and maintainability.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Compiling and testing are passed.

## Related PRs

https://github.com/facebook/docusaurus/pull/1997

### CODE REVIEW

File: packages/docusaurus-plugin-ideal-image/@types/react-ideal-image.d.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+declare module '@endiliey/react-ideal-image' {
```
Comment (yangshun) [inline packages/docusaurus-plugin-ideal-image/@types/react-ideal-image.d.ts]: We need to add the copyright header

```
/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
```

File: packages/docusaurus-plugin-ideal-image/src/index.ts [hunk 2]
```diff
@@ -4,10 +4,13 @@
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  */
+import {LoadContext} from '@docusaurus/types';
```
Comment (yangshun) [inline packages/docusaurus-plugin-ideal-image/src/index.ts]: Leave an empty line after the copyright header.

File: packages/docusaurus-plugin-ideal-image/src/theme/IdealImage.tsx [hunk 3]
```diff
@@ -5,9 +5,10 @@
  * LICENSE file in the root directory of this source tree.
  */
 import React from 'react';
+import {ImageProps} from '../types';
 import IdealImage from '@endiliey/react-ideal-image';
 
-function Image(props) {
+const Image: React.FC<ImageProps> = props => {
```
Comment (yangshun) [inline packages/docusaurus-plugin-ideal-image/src/theme/IdealImage.tsx]: TIL about `React.FC`.

But I prefer we write functional component as:

```
function Image(props: ImageProps): React.FC<ImageProps> {
  ...
}
```

File: packages/docusaurus-plugin-ideal-image/src/index.ts [hunk 4]
```diff
@@ -4,10 +4,13 @@
...
+import {PluginOptions} from './types';
+import {Configuration} from 'webpack';
 
-const path = require('path');
+import path from 'path';
 
-module.exports = function(context, options) {
+export = function(_context: LoadContext, options: PluginOptions) {
```
Comment (yangshun) [inline packages/docusaurus-plugin-ideal-image/src/index.ts]: I think this should be `export default`

**Comment by gimdongwoo** [comment]
@yangshun I’m really sorry. I am too busy recent 2 weeks.
I will be more attention for the next time.
Thank you so much.

**Comment by yangshun** [comment]
No problem, thanks for starting this, the fixes required were really minor. Thank you!!! 🎉

**Comment by endiliey** [comment]
OOps this PR broke something 
image

**Comment by yangshun** [comment]
It works when I build locally but not on Netlify. Super weird. I thought it
could be a caching issue so I uploaded my local files to Netlify lol.

**Pull request merged**